packages/loopring_v3/contracts/aux/bridge/Bridge.sol

```diff
@@ -0,0 +1,605 @@
+        WithdrawTransaction.Withdrawal memory withdrawal;
+        for (uint i = 0; i < tokens.length; i++) {
+            // Verify token data
+            require(
+                _getTokenID(tokens[i].token) == tokens[i].tokenID,
+                "INVALID_TOKEN_DATA"
+            );
+
+            // Verify withdrawal data
+            WithdrawTransaction.readTx(txsData, ctx.txIdx++ * ExchangeData.TX_DATA_AVAILABILITY_SIZE, withdrawal);
+            bytes20 onchainDataHash = WithdrawTransaction.hashOnchainData(
+                0,                  // Withdrawal needs to succeed no matter the gas coast
+                address(this),      // Withdraw to this contract first
+                new bytes(0)
+            );
+            require(
+                withdrawal.onchainDataHash == onchainDataHash &&
+                withdrawal.tokenID == tokens[i].tokenID &&
+                withdrawal.amount == withdrawalAmounts[i] &&
+                withdrawal.fee == 0,
+                "INVALID_BRIDGE_WITHDRAWAL_TX_DATA"
+            );
+
+            // Verify all tokens withdrawn were actually transferred into the Bridge
+            require(tokens[i].amount == 0, "INVALID_BRIDGE_TOKEN_DATA");
+        }
+    }
+
+    function _storeTransfers(InternalBridgeTransfer[] memory transfers)
+        internal
+    {
+        uint batchID = batchIDGenerator++;
+
+        // Store transfers to distribute at a later time
+        bytes32 hash = _hashTransfers(batchID, transfers);
+        require(pendingTransfers[hash] == 0, "DUPLICATE_BATCH");
+        pendingTransfers[hash] = block.timestamp;
+
+        // Log transfers to do
+        emit Transfers(batchID, transfers);
+    }
+
+    function _deposit(
+        address from,
+        address token,
+        uint96  amount
+        )
+        internal
+    {
+        if(amount > 0) {
+            // Do the token transfer directly to the exchange
+            uint ethValue = (token == address(0)) ? amount : 0;
+            exchange.deposit{value: ethValue}(from, address(this), token, amount, new bytes(0));
+        }
+    }
+
+    function _connectorCall(ConnectorCalls memory connectorCalls)
+        internal
+    {
+        // Check if the minimum amount of gas required is achieved
+        try IBridgeConnector(connectorCalls.connector).getMinGasLimit{gas: GAS_LIMIT_CHECK_GAS_LIMIT}(connectorCalls) returns (uint minGasLimit) {
+            require(connectorCalls.gasLimit >= minGasLimit, "GAS_LIMIT_TOO_LOW");
+        } catch {
+            // If the call failed for some reason just continue.
+        }
+
+        // Do the call
+        bool success;
+        try Bridge(this)._executeConnectorCall(connectorCalls) {
+            success = true;
+            emit BridgeCallSuccess(keccak256(abi.encode(connectorCalls)));
+        } catch Error(string memory reason) {
+            success = false;
+            emit BridgeCallFailed(keccak256(abi.encode(connectorCalls)), reason);
```

Comment:
make a variable for `keccak256(abi.encode(connectorCalls))`?

Reply:
Removed it for now and replaced it with just the connector address because `abi.encode(connectorCalls)` is pretty expensive. I think that should be fine, but will have to see what is cheap and useful.